src/frontend/Ast.ml

```diff
@@ -202,19 +213,27 @@ type typed_statement =
   statement_with
 [@@deriving sexp, compare, map, hash]
 
-let mk_typed_statement ~stmt ~loc ~return_type =
-  {stmt; smeta= {loc; return_type}}
+let mk_typed_statement ~stmt ~loc ~comments ~return_type =
+  {stmt; smeta= {loc; comments; return_type}}
 
 (** Program shapes, where we obtain types of programs if we substitute typed or untyped
     statements for 's *)
 type 's program =
-  { functionblock: 's list option
+  { comments0: comment list
+  ; functionblock: 's list option
+  ; comments1: comment list
   ; datablock: 's list option
+  ; comments2: comment list
   ; transformeddatablock: 's list option
+  ; comments3: comment list
   ; parametersblock: 's list option
+  ; comments4: comment list
   ; transformedparametersblock: 's list option
+  ; comments5: comment list
   ; modelblock: 's list option
-  ; generatedquantitiesblock: 's list option }
+  ; comments6: comment list
+  ; generatedquantitiesblock: 's list option
+  ; comments7: comment list }
```

Comment:
I think the naming here could be more clear - what do you think about something like
```
header_comment
functionblock
functionblock_comment
datablock
datablock_comment
  ...
```

Reply:
The naming definitely could be more clear. Those names are fine.